tests/lean/PPRoundtrip.lean.expected.out

```diff
@@ -11,21 +11,19 @@ List Nat
 List.{0} Nat
 id.{2} Nat
 Sum.{0, 0} Nat Nat
-id (@id Type Nat)
-fun (a : Nat) => a
-fun (a b : Nat) => a
-fun (a : Nat)
-  (b : Bool) => a
-fun {a b : Nat} => a
+id (id Nat)
+fun a => a
```

Comment:
This is because `knowsType` starts off with `true`, correct? Do you think the output is reasonable? Or should we make the initial value configurable and only set it to `true` in situations like `#check` where we show the expected type as well?

Reply:
I assume you are referring to the `fun a => a` line. Yes, this is because the top-down analyzer currently assumes that the original type is known, i.e. `knowsType := true`. I agree it may be nice to pass in `knowsType := false` in some circumstances, though the `checkM` function in this test seems a little artificial. The normal `check`/`print` both show the types:
```lean
#check fun (n : Nat) => n -- fun n => n : Nat → Nat

def foo := fun (n : Nat) => n
#print foo -- def foo : Nat → Nat := fun n => n
```
It is easy to expose the `knowsType` argument, the only question is whether it is worth the cost of weaving the argument all the way to the call sites.

Reply:
https://github.com/leanprover/lean4/commit/2da6847da45b7a7032e7c01b3ab2212565aadb14